sklearn/utils/extmath.py

```diff
@@ -863,20 +862,20 @@ def svd_flip(u, v, u_based_decision=True):
         Array v with adjusted rows and the same dimensions as v.
     """
     xp, _ = get_namespace(u, v)
-    device = getattr(u, "device", None)
+    u_device = device(u)
```

Comment:
Directly using `device` fails for dataframes. If `device = getattr(u, "device", None)` represent the semantics we want, we could modify the `device` function directly to do the same.